sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -458,15 +460,28 @@ impl<'vm, 'state, 'a> ExecutionContext<'vm, 'state, 'a> {
             The take+restore is an implementation detail of mutable references"
         );
         // restore is exclusively used for mut
-        let Ok((_, value_opt)) = self.borrow_mut_impl(arg, None) else {
+        let Ok((input_object_metadata, value_opt)) = self.borrow_mut_impl(arg, None) else {
             invariant_violation!("Should be able to borrow argument to restore it")
         };
-        let old_value = value_opt.replace(value);
-        assert_invariant!(
-            old_value.is_none() || old_value.unwrap().is_copyable(),
-            "Should never restore a non-taken value, unless it is copyable. \
-            The take+restore is an implementation detail of mutable references"
-        );
+        match input_object_metadata {
+            Some(InputObjectMetadata::Receiving{ id, version }) => {
+                let old_value = value_opt.replace(Value::Receiving(*id, *version));
+                // If the value is receiving and it's being restored it must have been taken since
+                // the `Receiving` struct is not copy.
+                assert_invariant!(
```

Comment:
restore is used to implement borrow_mut, so I think we will be able to hit this case

Reply:
I don't think we're able to hit this case since it doesn't have `copy` -- if we mutably borrow the value will be taken and then restored so `old_value` should always be none

Reply:
Ah, I see. Makes sense (just be sure to test it ;))